src/attached_probe.cpp

```diff
@@ -1029,7 +1032,7 @@ void AttachedProbe::attach_profile()
                                               PERF_COUNT_SW_CPU_CLOCK,
                                               period,
                                               freq,
-                                              pid,
+                                              pid == 0 ? -1 : pid,
```

Comment:
Wouldn't it make more sense to use -1 as the "no PID" value everywhere? Technically speaking, 0 is a valid PID.

Reply:
Soooo.... I started to do this but found out there is some weird behavior when using pid 0 at the attach point here (namely it doesn't seem to work - I've been doing some light poking around in the kernel to see what's going on). 
Basically, I agree with you but it would be mean utilizing the added AST filter approach for all these probes: hardware, software, interval, etc. instead of passing the pid at the attachpoint.

Reply:
Hm, there may be some misunderstanding here. I meant to change the semantics of `BPFtrace::pid()` to always return -1 when no PID is specified. Then, you can use it directly here (if 0 doesn't work as you claim) and you'll just need to update places which check for 0 as the "no PID" special case to check for -1.

Reply:
@viktormalik I understand but maybe I did a bad job of explaining the issue I ran into. So I started doing exactly as you described but when I set the pid to 0 on the command line e.g. `bpftrace -e 'hardware:cpu-cycles:10000000 { print(("hi", pid)); }' -p 0`. I was NOT getting events from only pid 0 (in fact, I wasn't getting any). However, if I did the filtering (via the added AST I added) instead of at the attach point for the hardware probe (which uses `bpf_attach_perf_event`) then it worked as expected.

Reply:
Ok, I'm still probably missing something. What did you expect to see from using `-p 0`? PID 0 is the swapper so it's not supposed to run any events so getting nothing is IMO actually correct.

The way I understand the semantics of `bpf_attach_perf_event` is:
- `pid == -1` -> trace all events system-wide
- `pid != -1` -> trace only events from the given PID

Reply:
Ah, I see, my mistake.

Still, the current solution is not correct, is it? When someone wants to filter events from PID 0 (i.e. uses `-p 0`), we will pass -1 here so all events will be captured, won't they? So users still need to apply manual filtering as you did in the above case.

But it's probably still better than my suggestion as we'd still need these ternary operators to replace 0 by -1 to at least get some events.

Reply:
> Still, the current solution is not correct, is it?

Yes, I'm still trying to figure out the best solution. Maybe we just don't pass the pid at the attach point for these other probes (until i have more time to investigate) and utilize the injected filter (as I'm doing for kfuncs/fentry) as that seems to yield more correct results. 

I still agree about allowing pid `-p` to be 0, which requires some additional changes.

Reply:
@viktormalik I'm going to look into allowing pid 0 in a follow-up PR as I need to think on it a bit more. After some discussion the use case for filtering on the swapper thread seems a little thin.

Reply:
For the internal APIs, I would recommend just using `std::optional<int>` for the arguments and plumbing. This avoids all the ambiguity over what the special sentinel value is. It seems reasonably fine to apply some constraints on this pid value at the top-level, but this shouldn't bring special consideration to the internal plumbing, because `std::optional` is using the type system to do the lifting :).